paywall/src/data-iframe/start/makePurchaseCallback.js

```diff
@@ -0,0 +1,44 @@
+import { getLocks, getTransactions, getKeys } from '../cacheHandler'
+
+// extraTip is not implemented yet, we need to do bignumber math, so post-pone this until walletService adds support
+const makePurchaseCallback = ({
+  walletService,
+  web3Service,
+  requiredConfirmations,
+  update,
+  window,
+}) =>
+  async function purchase(lockAddress /*, extraTip */) {
+    const [
+      { purchaseKey, processKeyPurchaseTransactions },
+      locks,
+      startingTransactions,
+      keys,
+    ] = await Promise.all([
+      import('../blockchainHandler/purchaseKey'),
+      getLocks(window),
+      getTransactions(window),
+      getKeys(window),
+    ])
+    const startingKey = keys[lockAddress]
+
+    Promise.all([
+      purchaseKey({
+        walletService,
+        lockAddress,
+        amountToSend: locks[lockAddress].keyPrice,
+      }),
+      processKeyPurchaseTransactions({
```

Comment:
I have to say that generally I am quite confused by the approach we've taken on these. It is strange that we call `processKeyPurchaseTransactions` to monitor a transaction which another promise is supposed to trigger.

I know this is a larger theme than this PR and I should have raised earlier, but what happens if `purchaseKey` fails?

Reply:
if purchaseKey fails, the promise throws, and throws out of the purchase callback. Then, we can do a simple try/catch and pass the error to the UI

Reply:
I have pushed a commit with the comment, and also 2 new tests for showing that any failure propagates out of the function.

Regarding the larger question, the approach taken is exactly the same as having walletService trigger the transaction, and web3Middleware start monitoring it through web3Service. Which is to say 2 separate entities, one does the writing, and the other monitors the transaction. The biggest difference is that the code here is written in a way that the logic flows from the top of the file (or function) to the bottom in chronological order, whereas the redux/middleware approach is entirely event-based.

Reply:
note to self: add lots of comments